File: dispatch-service/src/main/java/org/gsginzburg/dispatch/service/TenantService.java
/*
 * Copyright 2026 Gary Ginzburg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gsginzburg.dispatch.service;

import java.util.List;
import java.util.UUID;

import lombok.RequiredArgsConstructor;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import org.gsginzburg.dispatch.domain.dto.AssignTenantUserRequest;
import org.gsginzburg.dispatch.domain.dto.CreateTenantRequest;
import org.gsginzburg.dispatch.domain.model.Cluster;
import org.gsginzburg.dispatch.domain.model.Tenant;
import org.gsginzburg.dispatch.domain.model.TenantStatus;
import org.gsginzburg.dispatch.domain.model.TenantUser;
import org.gsginzburg.dispatch.domain.model.TenantUserId;
import org.gsginzburg.dispatch.domain.model.UserStatus;
import org.gsginzburg.dispatch.domain.repository.ClusterRepository;
import org.gsginzburg.dispatch.domain.repository.TenantRepository;
import org.gsginzburg.dispatch.domain.repository.TenantUserRepository;
import org.gsginzburg.shared.exception.TenantNotFoundException;

@Service
@RequiredArgsConstructor
public class TenantService {

    private final TenantRepository tenantRepository;
    private final ClusterRepository clusterRepository;
    private final TenantUserRepository tenantUserRepository;

    @Transactional(readOnly = true)
    public Page<Tenant> getTenants(int page, int size) {
        return tenantRepository.findAll(PageRequest.of(page, size, Sort.by("name")));
    }

    @Transactional(readOnly = true)
    public Tenant getTenant(UUID id) {
        return findTenant(id);
    }

    @Transactional
    public Tenant createTenant(CreateTenantRequest request) {
        Cluster cluster = clusterRepository.findById(request.getClusterId())
                .orElseThrow(() -> new RuntimeException("Cluster not found: " + request.getClusterId()));
        Tenant tenant = Tenant.builder()
                .name(request.getName())
                .cluster(cluster)
                .build();
        return tenantRepository.save(tenant);
    }

    @Transactional
    public Tenant updateTenantStatus(UUID id, TenantStatus status) {
        Tenant tenant = findTenant(id);
        tenant.setStatus(status);
        return tenantRepository.save(tenant);
    }

    @Transactional
    public Tenant assignCluster(UUID id, UUID clusterId) {
        Tenant tenant = findTenant(id);
        Cluster cluster = clusterRepository.findById(clusterId)
                .orElseThrow(() -> new RuntimeException("Cluster not found: " + clusterId));
        tenant.setCluster(cluster);
        return tenantRepository.save(tenant);
    }

    @Transactional
    public void deleteTenant(UUID id) {
        Tenant tenant = findTenant(id);
        tenant.setStatus(TenantStatus.ARCHIVED);
        tenantRepository.save(tenant);
    }

    @Transactional(readOnly = true)
    public List<TenantUser> getTenantUsers(UUID tenantId) {
        return tenantUserRepository.findByTenantId(tenantId);
    }

    @Transactional
    public void assignUser(UUID tenantId, AssignTenantUserRequest request) {
        TenantUser tu = TenantUser.builder()
                .tenantId(tenantId)
                .userId(request.getUserId())
                .role(request.getRole())
                .build();
        tenantUserRepository.save(tu);
    }

    @Transactional
    public void removeUser(UUID tenantId, UUID userId) {
        TenantUser tu = tenantUserRepository.findById(new TenantUserId(tenantId, userId))
                .orElseThrow(() -> new RuntimeException("User not in tenant"));
        tu.setStatus(UserStatus.INACTIVE);
        tenantUserRepository.save(tu);
    }

    private Tenant findTenant(UUID id) {
        return tenantRepository.findById(id)
                .orElseThrow(() -> new TenantNotFoundException(id.toString()));
    }
}

File: ui-dispatch/src/app/core/services/tenant.service.ts
/*
 * Copyright 2026 Gary Ginzburg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ApiResponse, PageResult, Tenant, TenantMembership } from '../models/api.model';

@Injectable({ providedIn: 'root' })
export class TenantService {
  private base = `${environment.apiUrl}/api/backoffice/tenants`;

  constructor(private http: HttpClient) {}

  getTenants(page = 0, size = 20): Observable<PageResult<Tenant>> {
    return this.http
      .get<ApiResponse<PageResult<Tenant>>>(this.base, {
        params: new HttpParams().set('page', page).set('size', size)
      })
      .pipe(map(r => r.data));
  }

  getTenant(id: string): Observable<Tenant> {
    return this.http
      .get<ApiResponse<Tenant>>(`${this.base}/${id}`)
      .pipe(map(r => r.data));
  }

  createTenant(name: string, clusterId: string): Observable<Tenant> {
    return this.http
      .post<ApiResponse<Tenant>>(this.base, { name, clusterId })
      .pipe(map(r => r.data));
  }

  updateStatus(id: string, status: string): Observable<Tenant> {
    return this.http
      .put<ApiResponse<Tenant>>(`${this.base}/${id}/status`, null, {
        params: new HttpParams().set('status', status)
      })
      .pipe(map(r => r.data));
  }

  assignCluster(id: string, clusterId: string): Observable<Tenant> {
    return this.http
      .put<ApiResponse<Tenant>>(`${this.base}/${id}/cluster`, null, {
        params: new HttpParams().set('clusterId', clusterId)
      })
      .pipe(map(r => r.data));
  }

  deleteTenant(id: string): Observable<void> {
    return this.http
      .delete<ApiResponse<void>>(`${this.base}/${id}`)
      .pipe(map(() => undefined));
  }

  getTenantUsers(tenantId: string): Observable<TenantMembership[]> {
    return this.http
      .get<ApiResponse<TenantMembership[]>>(`${this.base}/${tenantId}/users`)
      .pipe(map(r => r.data));
  }

  assignUser(tenantId: string, userId: string, role: string): Observable<void> {
    return this.http
      .post<ApiResponse<void>>(`${this.base}/${tenantId}/users`, { userId, role })
      .pipe(map(() => undefined));
  }

  removeUser(tenantId: string, userId: string): Observable<void> {
    return this.http
      .delete<ApiResponse<void>>(`${this.base}/${tenantId}/users/${userId}`)
      .pipe(map(() => undefined));
  }
}
